Adding CNV chart with CR and Operator

jordigilh opened the pull request:
Combines the CNV subscription and hyperconverged CR into a single chart. The chart contains a job whose responsibility is to delay the deployment of the hyperconverged CR until the CSV has reached the Successful state. This is to avoid the situation where the CR is created before the CRD.

@sabre1041 @nasx please review at your discretion.

sabre1041 commented:
@jordigilh we might be able to omit the need for a job by using this configuration in the `resourceCustomizations` section of the `ArgoCD` resource

```
    operators.coreos.com/Subscription:
      health.lua: |
        hs = {}
        if obj.status ~= nil then
          if obj.status.currentCSV ~= nil and (obj.status.state == "AtLatestKnown" or obj.status.state == "UpgradeAvailable" or obj.status.state == "UpgradePending") then
            hs.status = "Healthy"
            hs.message = "Subcription installed"
            return hs
          end
        end
        hs.status = "Progressing"
        hs.message = "Waiting for Subscription to complete."
        return hs
```

jordigilh commented:
> @jordigilh we might be able to omit the need for a job by using this configuration in the `resourceCustomizations` section of the `ArgoCD` resource
> 
> ```
>     operators.coreos.com/Subscription:
>       health.lua: |
>         hs = {}
>         if obj.status ~= nil then
>           if obj.status.currentCSV ~= nil and (obj.status.state == "AtLatestKnown" or obj.status.state == "UpgradeAvailable" or obj.status.state == "UpgradePending") then
>             hs.status = "Healthy"
>             hs.message = "Subcription installed"
>             return hs
>           end
>         end
>         hs.status = "Progressing"
>         hs.message = "Waiting for Subscription to complete."
>         return hs
> ```

Works great! thanks!. I've removed the job and it's resources and updated the [argocd.yaml](https://github.com/redhat-cop/rhel-edge-automation-arch/blob/653fbc49aa44b30286152cf57c240b3c5eaec518/openshift/gitops/manifests/bootstrap/argocd/base/argocd.yaml#L40-L52) file to include the `resourceCustomizations`.

I've tested it without the helm pre-hooks and works fine as well. So they're gone :smile:

sabre1041 commented on openshift/gitops/manifests/bootstrap/argocd/base/argocd.yaml:
```diff
@@ -37,3 +37,17 @@ spec:
+        if obj.status ~= nil then
+          if obj.status.currentCSV ~= nil and (obj.status.state == "AtLatestKnown" or obj.status.state == "UpgradeAvailable" or obj.status.state == "UpgradePending") then
+            hs.status = "Healthy"
+            hs.message = "Subcription installed"
```
```suggestion
            hs.message = "Subscription installed"
```

jordigilh replied:
Updated.

sabre1041 commented on charts/cnv/templates/subscription.yaml:
```diff
@@ -0,0 +1,10 @@
+kind: Subscription
+metadata:
+  name: hco-operatorhub
+  namespace: {{ .Values.namespace }}
```
` namespace: {{ template "common.names.namespace" $ }}`

jordigilh replied:
Done and removed the `namespace` field in the `values.yaml` file.

sabre1041 commented on charts/cnv/templates/operatorgroup.yaml:
```diff
@@ -0,0 +1,8 @@
+kind: OperatorGroup
+metadata:
+  name: kubevirt-hyperconverged-group
+  namespace: {{ .Values.namespace }}
```
` namespace: {{ template "common.names.namespace" $ }}

jordigilh replied:
Done

sabre1041 commented on charts/cnv/templates/kubevirt-hyperconverged.yaml:
```diff
@@ -1,12 +1,13 @@
-    deployOVS: {{ .Values.cnv.deploy_ovs | quote }}
   name: kubevirt-hyperconverged
-  namespace: {{ template "common.names.namespace" $ }}
+  namespace: {{ .Values.namespace }}
```
`namespace: {{ template "common.names.namespace" $ }}`

jordigilh replied:
done

sabre1041 commented:
@jordigilh This is really exciting! It worked great. we most likely want to add retries, but that can come later. a few small minor changes

sabre1041 reviewed (approved):
LGTM!
